### PR TITLE
bpo-37257:  obmalloc:  stop simple arena thrashing

### PR DESCRIPTION
**Pull request by tim-one**
[bpo-37257](https://bugs.python.org/issue37257):  obmalloc:  stop simple arena thrashing

https://bugs.python.org/issue37257

### CODE REVIEW

File: Objects/obmalloc.c [hunk 1]
```diff
@@ -1758,7 +1758,12 @@ pymalloc_free(void *ctx, void *p)
     /* All the rest is arena management.  We just freed
      * a pool, and there are 4 cases for arena mgmt:
      * 1. If all the pools are free, return the arena to
-     *    the system free().
+     *    the system free().  Except if this is the last
+     *    arena in the list, keep it to avoid trashing:
```
Comment (methane) [inline Objects/obmalloc.c]:
```suggestion
     *    arena in the list, keep it to avoid thrashing:
```

File: Objects/obmalloc.c [hunk 2]
```diff
@@ -1767,7 +1772,7 @@ pymalloc_free(void *ctx, void *p)
      *    nfreepools.
      * 4. Else there's nothing more to do.
      */
-    if (nf == ao->ntotalpools) {
+    if (nf == ao->ntotalpools && ao->nextarena!= NULL) {
```
Comment (methane) [inline Objects/obmalloc.c]:
```suggestion
    if (nf == ao->ntotalpools && ao->nextarena != NULL) {
```

**Comment by methane** [comment]
LGTM except two nitpicks.

File: Objects/obmalloc.c (hunk 2 above)
Comment (tim-one) [reply Objects/obmalloc.c]: @methane, thanks for that!  Those are kinds of silly typos I usually don't make .  Gladly accepted the changes :smile:.

**Pull request merged**